### ISSUE TITLE
XmHTML Build Instructions

### ISSUE THREAD

**Issue by TheChymera** [afni/afni#254]
Currently, there seems to be a fix for XmHTML which hinges on setting the build type, `set(CMAKE_BUILD_TYPE "Release")`. This means that the build behaviour can't be controlled at the top level, and if this is attempted by sed-ing out such separate control, the underlying bug is exposed.

From the [complete build log](https://chymera.eu/ppb/eff946.log):
```
[228/1925] : && /usr/bin/x86_64-pc-linux-gnu-gcc -fPIC -pthread -O3 -pipe  -Wl,-O1 -Wl,--as-needed   -Wl,--as-needed -Wl,--no-undefined -Wl,--no-as-needed -shared -Wl,-soname,libXmHTML.so -o targets_built/libXmHTML.so src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/HTTP.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/LZWStream.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/StringUtil.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib
/common/callbacks.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/colors.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/cookie.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/error.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/events.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/fonts.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/format.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/hash.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/images.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/layout.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/map.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/paint.c.o src/XmHTML/CMakeF
iles/XmHTML.dir/lib/common/parse.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/private.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/psoutput.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/public.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/quantize.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readBitmap.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readFLG.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readGIF.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readGIFplc.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readJPEG.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readJPEGplc.c.o src/XmHTML/CMakeFiles/XmHTM
L.dir/lib/common/readPNG.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readXPM.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/stack.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/Balloon.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/XmHTML.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/debug_menu.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/frames.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/motif.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/plc.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/textsel.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/XCC.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/XmImage.c.o src/XmHTML/
CMakeFiles/XmHTML.dir/lib/Motif/forms.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/output.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/strings.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/warnings.c.o  -Wl,-rpath,::::::::::::::::  /usr/lib64/libXm.so  /usr/lib64/libXt.so  /usr/lib64/libXpm.so  /usr/lib64/libXext.so  /usr/lib64/libX11.so  /usr/lib64/libjpeg.so  /usr/lib64/libICE.so  /usr/lib64/libSM.so && :         FAILED: targets_built/libXmHTML.so                                                                                                                                                                                  
: && /usr/bin/x86_64-pc-linux-gnu-gcc -fPIC -pthread -O3 -pipe  -Wl,-O1 -Wl,--as-needed   -Wl,--as-needed -Wl,--no-undefined -Wl,--no-as-needed -shared -Wl,-soname,libXmHTML.so -o targets_built/libXmHTML.so src/X
mHTML/CMakeFiles/XmHTML.dir/lib/common/HTTP.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/LZWStream.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/StringUtil.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/cal
lbacks.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/colors.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/cookie.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/error.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/
common/events.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/fonts.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/format.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/hash.c.o src/XmHTML/CMakeFiles/XmHTML.dir
/lib/common/images.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/layout.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/map.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/paint.c.o src/XmHTML/CMakeFiles/XmHTML
.dir/lib/common/parse.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/private.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/psoutput.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/public.c.o src/XmHTML/CMakeFi
les/XmHTML.dir/lib/common/quantize.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readBitmap.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readFLG.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readGIF.c.o sr
c/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readGIFplc.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readJPEG.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readJPEGplc.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/c
ommon/readPNG.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/readXPM.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/stack.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/Balloon.c.o src/XmHTML/CMakeFiles/XmHTML.
dir/lib/Motif/XmHTML.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/debug_menu.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/frames.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/motif.c.o src/XmHTML/CMakeFiles/
XmHTML.dir/lib/Motif/plc.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/textsel.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/XCC.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/XmImage.c.o src/XmHTML/CMakeFiles/
XmHTML.dir/lib/Motif/forms.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/output.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/strings.c.o src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/warnings.c.o  -Wl,-rpath,::::
::::::::::::  /usr/lib64/libXm.so  /usr/lib64/libXt.so  /usr/lib64/libXpm.so  /usr/lib64/libXext.so  /usr/lib64/libX11.so  /usr/lib64/libjpeg.so  /usr/lib64/libICE.so  /usr/lib64/libSM.so && :
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: src/XmHTML/CMakeFiles/XmHTML.dir/lib/common/LZWStream.c.o: in function `LZWStreamInit':                                             
LZWStream.c:(.text+0xd39): warning: the use of `tmpnam' is dangerous, better use `mkstemp'                
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/debug_menu.c.o: in function `debugCB':
debug_menu.c:(.text+0x6): undefined reference to `__rsd__debug_levels_defined'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: src/XmHTML/CMakeFiles/XmHTML.dir/lib/Motif/debug_menu.c.o: in function `_XmHTMLAddDebugMenu':                                       
debug_menu.c:(.text+0x2b8): undefined reference to `__rsd__debug_levels_defined'                                                                                                                                    
collect2: error: ld returned 1 exit status  
```

Is there an easy way of modifying the XmHTML files to avoid this bug across different compiler configurations?
@afni-rwcox @leej3

**Comment by leej3** [comment]
@afni-rwcox I remember that we took this path to avoid the breakage compiler flags are used. Who would have thought it would come back to bite us :) I can’t remember much else except that the different compiler flags used for the different build modes (release, debug) caused bits of the sources to be included that we didn’t want. 

It’s not the end of the world if it can’t be fixed by ifdef masking of the appropriate symbols but if it’s not too much effort to make the build more robust across different levels of compiler optimization that would be great. I imagine this will pop up for a lot of package managers that try to control the flags used by the compiler...

**Comment by leej3** [comment]
Specifically, the issue is that the Xmhtml build type is hard-coded in [this line](https://github.com/afni/afni/blob/c99cd25653baac1e0e4a0e3e6d18d22b5fb88f98/src/XmHTML/CMakeLists.txt#L4). This hard-coding is required or the above missing symbol error occurs. The portage package managing system on gentoo linux has some automated editing of build systems that attempts to reverse this such hard-coding, which would result in certain compiler flags being used that the portage system would have no control over.

For info in the different build types you can  (see [here](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html) for info on the cmake build types)

**Comment by afni-rwcox** [comment]
Well, that portage/gentoo choice sounds weird, presuming that it knows better how to build something than the person who designed the build. And then leaving it them to pick up the broken pieces. “It’s your fault for putting that fragile vase out where I knocked it to the brick floor.”

It seems like the flag -DNDEBUG is needed in the build. Can it be forced in by setting CMAKE_C_FLAGS instead of CMAKE_C_FLAGS_RELEASE ? Or are there other flags getting munged around?

On 02/27/2021, 11:07 AM, "john lee" <notifications@github.com> wrote:


Specifically, the issue is that the Xmhtml build type is hard-coded in this line<https://github.com/afni/afni/blob/c99cd25653baac1e0e4a0e3e6d18d22b5fb88f98/src/XmHTML/CMakeLists.txt#L4>. This hard-coding is required or the above missing symbol error occurs. The portage package managing system on gentoo linux has some automated editing of build systems that attempts to reverse this such hard-coding, which would result in certain compiler flags being used that the portage system would have no control over.

For info in the different build types you can (see here<https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html> for info on the cmake build types)

**Comment by leej3** [comment]
I think "NDEBUG" is it! Thank you. Typically that is not set for "release" builds and previously I had added it in explicitly (and then hard-coded the XmHTML build type to "release"). I'm not certain CMAKE_C_FLAGS will always propagate if the build setting is changed. I think I'll just add NDEBUG to all XmHTML targets regardless of build type since the build is broken if it is not set. I have done that [here](https://github.com/leej3/afni/commit/c54721b6125c09934da8ff33bc74c9cfd334142d). If that's ok with you and we've confirmed it works with the portage build (and it doesn't break other stuff) then this is sufficient to close the issue.

Consider the vase glued back together!

> Well, that portage/gentoo choice sounds weird, presuming that it knows better how to build something than the person who designed the build

It's not so much about presuming to know better. During packaging there will typically be trade offs for the various build flags for  performance/security. Such tuning is often done en masse for the purposes of binary interface compatibility across all the different software projects that are packaged.These tweaks occur on a OS or package manager specific basis (and likely changes over time). This fix will hopefully be more robust across the different contexts.

**Comment by leej3** [comment]
I confirm that this solution fixes the issue for the portage build and breaks none of the CI tests for AFNI. Fix is currently on #241. That will be merged once we add some tweaks to enforce .so versions on nifti,gifti, and gts builds.

**Comment by afni-rwcox** [comment]
As far as I’m concerned, if we ever have to debug XmHTML, I would just throw it away and do without it somehow. In other words, NDEBUG is OK (I suppose the ‘N’ means ‘No’).



On 02/27/2021, 2:22 PM, "john lee" <notifications@github.com> wrote:


I think "NDEBUG" is it! Thank you. Typically that is not set for "release" builds and previously I had added it in explicitly (and then hard-coded the XmHTML build type to "release"). I'm not certain CMAKE_C_FLAGS will always propagate if the build setting is changed. I think I'll just add NDEBUG to all XmHTML targets regardless of build type since the build is broken if it is not set. I have done that here<https://github.com/leej3/afni/commit/c54721b6125c09934da8ff33bc74c9cfd334142d>. If that's ok with you and we've confirmed it works with the portage build (and it doesn't break other stuff) then this is sufficient to close the issue.

Consider the vase glued back together!

Well, that portage/gentoo choice sounds weird, presuming that it knows better how to build something than the person who designed the build

It's not so much about presuming to know better. During packaging there will typically be trade offs for the various build flags for performance/security. Such tuning is often done en masse for the purposes of binary interface compatibility across all the different software projects that are packaged.These tweaks occur on a OS or package manager specific basis (and likely changes over time). This fix will hopefully be more robust across the different contexts.
